### ISSUE TITLE
Feature impact - for better or for worse

### ISSUE THREAD

**Issue by ehudkr** [shap/shap#144]
Regarding the Tree-SHAP summary plot. It is a great way to assess the impact of features on the *model's* output, but one thing I'm missing is validating whether that impact improves the prediction or worsens it.

If I understand correctly from the paper, the SHAP values are calculated in equation (2):
![image](https://user-images.githubusercontent.com/15989012/42282632-d263b0fa-7faf-11e8-80eb-dfa2ccdc24b9.png)
Measuring the impact of feature *i* on the model's prediction.

I would like to argue that it misses an important component: the loss of the model with and without that added feature.
Lets denote *l(x,y)* as some loss function (lower the better), then for each feature *i* we keep two additional values:
![image](https://user-images.githubusercontent.com/15989012/42284166-51801e4c-7fb4-11e8-9dff-fea35c7c16bf.png)

We can now define two statistics:
![image](https://user-images.githubusercontent.com/15989012/42284216-7e2917f0-7fb4-11e8-9e2a-18e2fd9bbd75.png)
The latter, of course, used for the SHAP value. 

We can look at the 4 resulting possibilities:
```
Δf>0 & Δl>0    high impact, screws prediction
Δf>0 & Δl<0    high impact, benefits prediction
Δf<0 & Δl>0    low impact,  screws prediction
Δf<0 & Δl<0    low impact,  benefits prediction
```
We can then add another "dimension" to the summary plot describing `sign(Δl)`. One simple way that occurred to me is, circle-markers for Δl>0 and x-markers for Δl<0. Another way could be an optional coloring of the circles according to `Δl`, instead of the current way (that is, the value of the feature of the individual)

I would like to hear your take on this idea. Does it even sounds reasonable? Am I missing something obvious? 
Especially since I strictly assume there's no monotonicity of loss when adding features to a tree model. Which I believe to be a fair assumption, since all trees have the same model specification - mainly, they do not differ by their hyper-parameters. Therefore, the more features, the more complex the tree ought to be, and since models have constant complexity along the process, we expect the loss to vary.

**Comment by slundberg** [comment]
Interesting comments and questions! I am assuming "high impact" and "low impact" mean positive and negative impacts, not important and unimportant, correct me if that's not what you meant.

I think the general idea is interesting, though it could be challenging to compute. I thought briefly about extending the global SHAP plots to account for even just the simple squared loss, but it makes the Tree SHAP algorithm harder (and I didn't see a solution).

In practice, if you have something like a logistic loss on a sample with a label of 1, then every feature with a positive SHAP value will also likely have `Δl < 0` while every negative feature will likely have `Δl > 0`. If you have a practical example where this is different that would be interesting. I think we might be able to use some ideas from Deep SHAP (`DeepExplainer`) to estimate the loss function based SHAP values you are thinking of. (If we use conditional expectations to model missing features then the loss based formulation you are proposing is really the SHAP values of a new function, where the output is not the model output, but the loss output)

**Comment by maybeluo** [comment]
@slundberg logloss/squared-loss can't be added directly like weight and the fact makes them difficult to be calculated. 
But when we use model_output=`logloss`, the sum exactly to logloss. I'm confused here. If we just calculate logloss _l(x,y)_ individualy for each tree, the sum might not equal to the final logloss.
Could you explain what's the meaning of model_output=`logloss` and how to compute them? I couldn't figure it out from the docs of TreeExplainer.
